### PR TITLE
Added methods to the Enumerable Module

### PR DESCRIPTION
**Pull request by dannisonarias**
I have completed the recreation of some of the methods in the Enumerable Modules such as .each
.each_with_index and .inject.

### CODE REVIEW

**Review by fatymahmed — CHANGES_REQUESTED** [review]
### Hi @dannisonarias 
Great job on what you have done so far 👏👏, just a few changes and you are good to go.
- [ ] Make sure you have installed stickler on your repo, check [this](https://github.com/microverseinc/linters-config/tree/master/ruby) on how to do so, you have the config files but you missed a step (have you enabled stickler to your repo?)
- [ ] Please don't change stickler config as mentioned on Pathwright, [see screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/general/linter_config.png), the only exception is

> You can have Style/CaseEquality cop disabled as described [here](https://rubydoc.info/gems/rubocop/0.24.1/frames#Disabling_Cops_within_Source_Code) see [Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/style_case_equality.png) from Pathwright

- Some of your enumerable methods aren't working correctly, check my comments below for a detailed review on them. You can also check the enumerable methods in this [guide](https://ruby-doc.org/core-2.6.5/Enumerable.html#method-i-each_with_index) and all the different arguments passed to each of them.

### Status: Changes Required 🛑
_Feel free to reach out to me on slack (Fatima) if you have any questions, I'd be happy to help._

File: main.rb
Comment (fatymahmed) [inline main.rb]: - For my_each_with_index, it should call the given block once for each element in self.
According to the enumerable guide [here](https://ruby-doc.org/core-2.6.5/Enumerable.html#method-i-each_with_index), the item comes before the index, however, in your method, (line 23), the index comes before the item.

For example
`[1,2,3,4].my_each_with_index{ |num, idx| p "Num: #{num}, idx: #{idx}\n" }`
The built in method returns
`"Num: 0, idx: 1\n"
"Num: 1, idx: 2\n"
"Num: 2, idx: 3\n"
"Num: 3, idx: 4\n"
[1, 2, 3, 4]`

`[1,2,3,4].my_each_with_index{ |num, idx| p "Num: #{num}, idx: #{idx}\n" }`
Yours returns
`"Num: 1, idx: 0\n"
"Num: 2, idx: 1\n"
"Num: 3, idx: 2\n"
"Num: 4, idx: 3\n"
[1, 2, 3, 4]
`

File: main.rb
Comment (fatymahmed) [inline main.rb]: - For my_all? let's handle the case when no block or argument is given, it returns true when none of the collection members are false or nil.
  For example
> 
> The built-in method
> `[1, true, 'hi', [ ]].all?`
> returns `true`
> 
> Your method
> `[1, true, 'hi', [ ]].my_all?`
> returns `false`

File: main.rb
Comment (fatymahmed) [inline main.rb]: - Handle the case when a pattern other than Regex or a Class is given, it returns false if none of the collection matches the pattern
For example

The built-in method
`%w[cat door rod blade].any?('cat')
=> true
`

Your method
` %w[cat door rod blade].my_any?('cat')
=> false
`
- It should return true since one element in the array matches the pattern

File: main.rb
Comment (fatymahmed) [inline main.rb]: - Handle the case when a pattern other than Regex or a Class is given, it returns true only if none of the collection matches the pattern


For example
Built-in method
`[5, 'door', 'rod', 'blade'].none?(5)` returns `false`

Your method
`[5, 'door', 'rod', 'blade'].my_none?(5)` returns `true`
- It should return false since at least 1 element matches the pattern.

File: main.rb
Comment (fatymahmed) [inline main.rb]: - For my_map, handle the case when no block is given, it should return an Enumerator.

Built-in method
`[5, 'door', 'rod', 'blade'].map`
returns `#<Enumerator: [5, "door", "rod", "blade"]:map>
`


Your method
`[5, 'door', 'rod', 'blade'].my_map `
expects a block but when none if provided, it gives the following error
> LocalJumpError:
>   no block given (yield)

File: main.rb
Comment (fatymahmed) [inline main.rb]: - For my_inject handle the case when a symbol is specified, it combines each element of the collection by applying the symbol as a named method

 For example
 Built-in method
 `Range.new(5, 50).inject(2, :*)` returns    `2534507766809448170301050680505397403698136797413376000000000000
`

 Your method
 `Range.new(5, 50).my_inject(2, :*)` returns  `2`
- Make sure your method returns the same value as the built-in one.

**Comment by dannisonarias** [comment]
I've tried to fix some of the complexity issues but there just seems to be no way of satisfying that rule. If I create methods for my control flow,  i will not be able to stay within 100 lines of code. any suggestions?

**Comment by dannisonarias** [comment]
I can't find a reasonable solution to the rubocop complexity issues. I've tested some of the original methods and still get some errors. My module works! I've addressed previous issues and did the testing. I am ready to move on.

File: main.rb
Comment (stratospherique) [inline main.rb]: * [ ] `my_all?` is identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
  ```ruby
    array = [1,2,5,4,7]
    p array.my_all?(5) == array.all?(5) # false
  ```

File: main.rb
Comment (stratospherique) [inline main.rb]: * [ ] `my_any?` is identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class
  ```ruby
    array = [1,2,5,4,7]
    p array.my_any?(Integer) == array.any?(Integer) # false
  ```

File: main.rb
Comment (stratospherique) [inline main.rb]: * [ ] `my_inject` is identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block
  ```ruby
    array = [4,7,8,89,9,6,5,12,4,2,5]
    operation = proc { |sum, n| sum + n  }
    p array.my_inject(&operation) == array.inject(&operation) # false
  ```

* [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  ```ruby
    array = [1,7,8,89,9,6,5,12,4,2,5]
    p array.my_inject(:+) == array.inject(:+) # false
  ```

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Impressive work implementing the suggested changes :clap: However, there are methods that need further refactoring in order to get approval for the project. Check :arrow_down: 

- **[REQUIRED]** Check the code comments below, they should help you refactor your methods :gear: 

### That's it :relieved: When finished, kindly resubmit another code review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

**Comment by dannisonarias** [comment]
Changes requested by Stratospherique have been implemented.

File: main.rb
Comment (suretrust) [inline main.rb]: * [ ] should have working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5].my_none?{ |num| num > 9 } != [1,2,3,4,5].none?{ |num| num > 9 }
```

File: main.rb
Comment (suretrust) [inline main.rb]: * [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1, 2, 3, 4, 5, 6, 7, 8, 9].my_inject(2, :*) != [1, 2, 3, 4, 5, 6, 7, 8, 9].inject(2, :*)
```

**Review by suretrust — CHANGES_REQUESTED** [review]
Hi @dannisonarias, 👋 

### Great job thus far! 🙌 Just a few more cases to cover, please see my comments on your code. 👇 

_Happy coding!_ 🎉

File: main.rb
Comment (Forison) [inline main.rb]: - [ ] ``` multiply_els([2,4,5])``` should return``` 40``` currently, it does not

**Review by Forison — CHANGES_REQUESTED** [review]
## Status: Require changes🛑
# 🙋‍♂️  Danny
Good job on implementing the changes suggested in the previous review
Here is one thing we have to do to get this project approved
* [ ] should have working multiply_els  that uses my_inject as described on Odin ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/multiply_els.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
> From Odin: Test your #my_inject by creating a method called #multiply_els which multiplies all the elements of the array together by using #my_inject, e.g. multiply_els([2,4,5]) #=> 40
## That is it kindly do this change and submit project or review
#### Happy coding
#### Addo Forison

**Comment by Anna-Myzukina** [comment]
## Hi @dannisonarias !
### Your project looks good :+1: , however, there are a few things that you need to fix before it can be approved.

* Next methods my_none? and my_inject still needs changes:

- [ ] should have working my_none? identical and returns the same thing as ruby's [none?](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). 
from  [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)
screen
- [ ] should have working my_inject identical and returns the same thing as ruby's [inject](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)
Screenshot 











### Please kindly fix the above issues and send another code review request.
* If you have any questions or need help, you can message me on slack (@ Anna Muzykina). Happy coding

**Review by Anna-Myzukina — CHANGES_REQUESTED** [review]
Next methods: my_none? and my_inject still needs changes

**Comment by dannisonarias** [comment]
I've applied for another code review. I don't think this is a proper review of my code as I have nothing to work with @Anna-Myzukina

**Review by pelzolga123 — CHANGES_REQUESTED** [review]
Project requires changes.

**Comment by pelzolga123** [comment]
## Hi Dannison!  
### Your did a good job so far!
Like @Anna-Myzukina pointed, there are two methods, that require fixing:
* [ ] `my_none?`
* When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern:
`[1,2,3,4,5].my_none?(2)` - should return `false`
* [ ] `my_inject`
* Combines all elements of enum by applying a binary operation, specified by a block.
` (5..11).my_inject(4) { |i, n| i * n }` - should return `6652800`

### Please kindly fix the above issues and send another code review request.

If you have any questions or need help, you can message me on slack (@ Olga Pelts). *Happy coding*

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @dannisonarias :handshake: 

### Thank you for doing all the changes requested by the previous reviewers :medal_sports: , you have a few more changes to make as a follow up of the previous changes and you'll be good to go :construction: :construction_worker_man: 

## my_inject

- [ ] when a symbol or a string operator is specified combines each element of the collection by applying the symbol or string as a named method

`puts [5, 1, 2].my_inject("+")` should be equal to `puts [5, 1, 2].inject("+")` or `puts [5, 1, 2].my_inject(:+)` or `puts [5, 1, 2].inject(:+)`

`puts ["5,", "1,", "2,"].my_inject("+")` should be equal to `puts ["5,", "1,", "2,"].inject("+")` or `puts ["5,", "1,", "2,"].my_inject(:+)` or `puts ["5,", "1,", "2,"].inject(:+)`

- [ ] combines all elements of enum by applying a binary operation, specified by a proc

if given 

operation = proc { |sum, n| sum + n }

`[1,5,8].my_inject(&operation)` should be equal to `[1,5,8].inject(&operation)`

## my_map

#### should take a block or a proc or both

- [ ] [OPTIONAL] Modify your #my_map method to take a proc or a block or both when a proc and a block are given, only execute the proc. the real map doesn't have this functionality but the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) (step #12) requires you to add it. I'm sorry this wasn't pointed out by the previous reviewers but this is actually in the project specs so I deeply encourage you to have it.

given

`my_proc = Proc.new {|num| num > 10 }`

`puts [18, 20, 5, 1].my_map(my_proc) {|num| num < 10 } # should return true true false false`

## Optional

#### my_none

I think we could get the block from my_none? from my_any? we should only do something like !my_any? for my_none? to work. For example:

```
  def my_none?(arg = nil, &block)
    !my_any?(arg = nil,&block)
  end
```
To make this work we have to use block.call(element) instead of yield(element) in my_any?

### That's it! :raised_hands: 

### Please :pray: do the changes :man_technologist: and submit back for a code review when you are done :memo: 

Happy coding! :rocket: 

Any questions? ask me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :traffic_light:

**Comment by dannisonarias** [comment]
I am not going to make any more changes to the code that is unnecessary right now because this code wasn't written with OOD (because we haven't gotten to that yet), and it seems like every time I change one piece of code I need to refactor other things that break. Whenever I test a proc AND block with my_map I get the Error 'both block arg and actual block given'. I updated my_map to accept a proc.

**Comment by dannisonarias** [comment]
@euqueme
═✿✿✿═════✿✿═════✿✿═════✿✿✿═
════════════ ('\../') ═════════════
════════════ (◕.◕) ═════════════
════════════ (,,)(,,) ═════════════
.▀█▀.█▄█.█▀█.█▄.█.█▄▀　█▄█.█▀█.█─█
─.█.─█▀█.█▀█.█.▀█.█▀▄　─█.─█▄█.█▄█
for your very thorough review, I know it takes a lot of time but ultimately it saves a lot more time for both parties.

**Review by whiz25 — APPROVED** [review]
## Nice Progress :clap: 
You have made the changes required and that's good. There are two failing edge cases but I am approving this project since it has exceeded the max review cycles.
#### Fix the failing edge cases before merging. Check my comments for suggestions.

### Martin Afani

File: main.rb
Comment (whiz25) [inline main.rb]: - [ ] combines all elements of enum by applying a binary operation, specified by a block
```ruby
array = Array.new(100) { rand(0...9) } 
words = %w[dog door rod blade]
operation = proc { |sum, n| sum + n } 
search = proc { |memo, word| memo.length > word.length ? memo : word }

  array.my_inject(&operation) == array.inject(&operation) #true
  actual = range.my_inject(4) { |prod, n| prod * n }
  expected = range.inject(4) { |prod, n| prod * n }
  expect(actual) == expected #true
  words.my_inject(&search) == words.inject(&search) #true
```
- [ ] when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
range = Range.new(5, 50) 
array.my_inject(:+) == array.inject(:+) #true
range.my_inject(2, :*) == range.inject(2, :*) #true
```

**Pull request merged**
